### PR TITLE
Refactor config package, integrate new logging

### PR DESCRIPTION
**Pull request by ac55-sanger**

### CODE REVIEW

File: jobqueue/scheduler/kubernetes.go [hunk 1]
```diff
@@ -162,22 +161,17 @@ func (c *ConfigKubernetes) GetServerKeepTime() time.Duration {
...
 
-	s.Logger = logger.New("scheduler", "kubernetes")
 	kubeLogFile := filepath.Join(s.config.ManagerDir, kubeSchedulerLog)
-	fh, err := log15.FileHandler(kubeLogFile, log15.LogfmtFormat())
-	if err != nil {
-		return fmt.Errorf("wr kubernetes scheduler could not log to %s: %s", kubeLogFile, err)
-	}
 
-	l15h.AddHandler(s.Logger, fh)
+	clog.ToFileAtLevel(kubeLogFile, "warn")
```
Comment (sb10) [inline jobqueue/scheduler/kubernetes.go:169]: This is something we can't do, and is a problem with having a single global logger. We need the ability to have some logs going to one place, and other logs going to another place.
As a simple fix, just continue to use log15 directly here, so largely undo the change.

File: jobqueue/scheduler/local.go [hunk 2]
```diff
@@ -173,10 +173,8 @@ type job struct {
 
 // initialize finds out about the local machine. Compatible with amd64 archs
 // only!
-func (s *local) initialize(config interface{}, logger log15.Logger) error {
+func (s *local) initialize(ctx context.Context, config interface{}) error {
 	s.config = config.(*ConfigLocal)
-	s.Logger = logger.New("scheduler", "local")
```
Comment (sb10) [inline jobqueue/scheduler/local.go:178]: We'll need something like clog.ContextWithScheduler() to replace this.

File: jobqueue/scheduler/kubernetes.go [hunk 3]
```diff
@@ -308,58 +300,58 @@ func (s *k8s) reqCheck(req *Requirements) error {
...
 	s.es = resp.Ephemeral
 
 	return resp.Error
 }
 
 // setMessageCallBack sets the given callback function.
-func (s *k8s) setMessageCallBack(cb MessageCallBack) {
-	s.Debug("setMessageCallBack called")
+func (s *k8s) setMessageCallBack(ctx context.Context, cb MessageCallBack) {
+	clog.Debug(ctx, "setMessageCallBack called")
```
Comment (sb10) [inline jobqueue/scheduler/kubernetes.go:328]: These debug lines are unnecessary, and therefore so is changing the callback methods to take a ctx?

File: jobqueue/scheduler/scheduler.go [hunk 4]
```diff
@@ -223,7 +223,7 @@ type Scheduler struct {
 // Providing a logger allows for debug messages to be logged somewhere, along
```
Comment (sb10) [inline jobqueue/scheduler/scheduler.go:223]: Again, do we need some kind of optional logging to replace the loss of logger?

File: limiter/limiter.go [hunk 5]
```diff
@@ -33,7 +34,7 @@ import (
 // a group from some on-disk database, so you don't have to have all group
 // limits in memory. (Limiter itself will clear out unused groups from its own
 // memory.)
-type SetLimitCallback func(name string) int
+type SetLimitCallback func(context.Context, string) int
```
Comment (sb10) [inline limiter/limiter.go:37]: Why is context passed in to many methods of limiter now? I don't see it being used in the diff.

File: queue/queue.go [hunk 6]
```diff
@@ -1218,7 +1209,7 @@ func (queue *Queue) Remove(key string) error {
 // depending upon it. You'd want to check this before Remove()ing this item if
 // you're removing it because it was undesired as opposed to complete, as
 // Remove() always triggers dependent items to become ready.
-func (queue *Queue) HasDependents(key string) (bool, error) {
+func (queue *Queue) HasDependents(ctx context.Context, key string) (bool, error) {
```
Comment (sb10) [inline queue/queue.go:1212]: Why does this take a context?

**Review by sb10 — CHANGES_REQUESTED** [review]
Huge change, thanks for your hard work.

While we're now using a literal context logger, we've unfortunately lost the actual context in the log output, so it needs to be re-introduced.

File: jobqueue/scheduler/openstack.go [hunk 7]
```diff
@@ -827,12 +837,8 @@ func (s *opst) reqForSpawn(req *Requirements) *Requirements {
 
 // spawn creates a new instance in OpenStack. Errors are not returned but are
 // logged, and problematic servers are terminated.
-func (s *opst) spawn(req *Requirements, flavor *cloud.Flavor, requestedOS string, requestedScript []byte, requestedConfigFiles string, needsSharedDisk bool, cmd string, call string) {
-	// since we can have many simultaneous calls to this method running at once,
-	// we make a new logger with a unique "call" context key to keep track of
-	// which spawn call is doing what
-	logger := s.Logger.New("call", call, "flavor", flavor.Name)
```
Comment (sb10) [inline jobqueue/scheduler/openstack.go:834]: Instead of taking a call value, the caller of spawn should set call value in the ctx.

So it looks like we need new clog methods to add a unique call value, and a server flavour.

File: jobqueue/scheduler/openstack.go [hunk 8]
```diff
@@ -883,26 +889,25 @@ func (s *opst) spawn(req *Requirements, flavor *cloud.Flavor, requestedOS string
...
 	failMsg := "server failed spawn"
-	logger.Debug("will spawn new server", "cmd", cmd)
+
+	clog.Debug(ctx, "will spawn new server", "cmd", cmd)
+
 	tSpawn := time.Now()
-	server, err := s.provider.Spawn(requestedOS, osUser, flavor.ID, req.Disk, s.config.ServerKeepTime, false, usingQuotaCB)
-	serverID := "failed"
-	if server != nil {
-		serverID = server.ID
```
Comment (sb10) [inline jobqueue/scheduler/openstack.go:891]: Our ctx should get the server.ID (or "failed") added to it like this.

File: jobqueue/scheduler/openstack.go [hunk 9]
```diff
@@ -1072,10 +1077,8 @@ func (s *opst) cmdNotNeeded(cmd string) {
 // if we can't start the cmd, not if the command fails (schedule() only
 // guarantees that the cmds are run count times, not that they are /successful/
 // that many times).
-func (s *opst) runCmd(cmd string, req *Requirements, reservedCh chan bool, call string) error {
-	logger := s.Logger.New("call", call)
-
-	requestedOS, requestedScript, requestedConfigFiles, requestedFlavor, needsSharedDisk, err := s.serverReqs(req)
+func (s *opst) runCmd(ctx context.Context, cmd string, req *Requirements, reservedCh chan bool, call string) error {
+	requestedOS, requestedScript, requestedConfigFiles, requestedFlavor, needsSharedDisk, err := s.serverReqs(ctx, req)
```
Comment (sb10) [inline jobqueue/scheduler/openstack.go:1081]: Instead of taking a call value, the caller of runCmd should set call value in the ctx.

File: jobqueue/scheduler/scheduler.go [hunk 10]
```diff
@@ -248,34 +247,25 @@ func New(name string, config interface{}, logger ...log15.Logger) (*Scheduler, e
...
-	} else {
-		l = log15.New()
-		l.SetHandler(log15.DiscardHandler())
-	}
-	s.Logger = l
-
 	s.Name = name
 	s.limiter = make(map[string]int)
-	err := s.impl.initialize(config, l)
+	err := s.impl.initialize(ctx, config)
```
Comment (sb10) [inline jobqueue/scheduler/scheduler.go:252]: For all of the `impl.method(ctx, ...)` calls, we should pass s.typeContext(), where that returns `clog.ContextWithSchedulerType(s.Name)`.

File: jobqueue/scheduler/scheduler_test.go [hunk 11]
```diff
@@ -52,27 +52,28 @@ func init() {
 }
 
 func TestLocal(t *testing.T) {
+	ctx := context.Background()
```
Comment (sb10) [inline jobqueue/scheduler/scheduler_test.go:55]: Somewhere there should be a test that our logs include the schedulertype context.

**Review by sb10 — CHANGES_REQUESTED** [review]

File: cloud/server.go [hunk 12]
```diff
@@ -516,7 +520,7 @@ func (s *Server) SSHClient(ctx context.Context) (*ssh.Client, int, error) {
 
 	// create a new client, add it to the pool
 	if s.sshClientConfig == nil {
-		err := s.createSSHClientConfig()
+		err := s.createSSHClientConfig(ctx)
```
[the remark this answers is not in the document]
Comment (ac55-sanger) [reply cloud/server.go:523]: because the context here is already of clog.ContextWithServerID() type. 
All the methods calling SSHClient has their context of clog.ContextWithServerID() type.

File: cloud/cloud.go [hunk 13]
```diff
@@ -757,7 +758,7 @@ func (p *Provider) CheckServer(serverID string) (working bool, err error) {
 		if _, present := p.resources.Servers[serverID]; present {
 			delete(p.resources.Servers, serverID)
 			p.Unlock()
-			err = p.saveResources()
+			err = p.saveResources(p.cloudContext(ctx))
```
Comment (sb10) [inline cloud/cloud.go:761]: Some saveResources() calls you call cloudContext(), some you don't. I think none of them need it because saveResources() itself calls cloudContext(). Double-check and make consistent for all calls.

File: jobqueue/scheduler/openstack.go [hunk 14]
```diff
@@ -1133,23 +1147,24 @@ func (s *opst) runCmd(cmd string, req *Requirements, reservedCh chan bool, call
...
-		err = s.local.runCmd(cmd, req, reserved, call)
+
+		err = s.local.runCmd(ctx, cmd, req, reserved)
 	} else {
 		if s.config.Umask > 0 {
 			cmd = fmt.Sprintf("(umask %d && %s)", s.config.Umask, cmd)
 		}
-		logger.Debug("running command remotely", "cmd", cmd)
+		clog.Debug(ctx, "running command remotely", "cmd", cmd)
 		_, _, err = server.RunCmd(context.Background(), cmd, false)
```
Comment (sb10) [inline jobqueue/scheduler/openstack.go:1168]: Make sure there's no use of context.Background() in package code; pass the ctx

**Review by sb10 — CHANGES_REQUESTED** [review]
Still some unresolved questions from previous review as well.

File: cloud/server.go [hunk 15]
```diff
@@ -1299,7 +1327,8 @@ func (s *Server) Alive(checkSSH ...bool) bool {
 		if err != nil {
 			return false
 		}
-		s.CloseSSHSession(session, clientIndex)
+
+		s.CloseSSHSession(ctx, session, clientIndex)
```
[the remark this answers is not in the document]
Comment (ac55-sanger) [reply cloud/server.go:1332]: Even if we call it more than once, it doesn't get duplicated in the log output. I tested this in clog package, it just updates the existing context with the latest serverid.

File: cmd/manager.go [hunk 16]
```diff
@@ -575,7 +578,7 @@ func logStarted(s *jobqueue.ServerInfo, token []byte) {
 
 	// go back to just stderr so we don't log token to file (this doesn't affect
 	// server logging)
-	appLogger.SetHandler(log15.LvlFilterHandler(log15.LvlInfo, log15.StderrHandler))
+	clog.ToDefaultAtLevel("info")
```
[the remark this answers is not in the document]
Comment (ac55-sanger) [reply cmd/manager.go:586]: Yes, clog.ToDefaultAtLevel sets the global logger to log to STDERR at the given level which is what appLogger was doing.

[the remark this answers is not in the document]
Comment (sb10) [reply cmd/manager.go:586]: The original code is changing what the CLI logs to, while there was a separate logger used within the Server that logs to a file. The question is, does the server continue to log to the file after we call `clog.ToDefaultAtLevel("info")`?

I'm planning on testing this manually when I get some time.

[the remark this answers is not in the document]
Comment (ac55-sanger) [reply cmd/manager.go:588]: Tested it on server. It doesn't affect the server logging.

[the remark this answers is not in the document]
Comment (ac55-sanger) [reply cmd/manager.go:588]: No, it doesn't work as expected. The 'log' file under .wr_<deployment> directory stays empty on the manager.

**Review by sb10 — CHANGES_REQUESTED** [review]
We need server info logs about starting and stopping to go to the log file when not in --debug mode. We also need the CLI to format it's logs like before.

File: cmd/add.go [hunk 17]
```diff
@@ -671,7 +673,7 @@ func parseCmdFile(jq *jobqueue.Client, diskSet bool) ([]*jobqueue.Job, bool, boo
 // path based on the file's MD5, and then returns an altered input value to use
 // the MD5 paths as the sources, keeping the desired destinations. It does not
 // alter path specs for config files that don't exist locally.
-func copyCloudConfigFiles(jq *jobqueue.Client, configFiles string) string {
+func copyCloudConfigFiles(ctx context.Context, jq *jobqueue.Client, configFiles string) string {
```
Comment (sb10) [inline cmd/add.go:676]: This ctx is not used?

File: cmd/manager.go [hunk 18]
```diff
@@ -735,23 +747,24 @@ func startJQ(postCreation []byte) {
 	sync.Opts.OnPotentialDeadlock = func() {
 		wgMsg := wgDebug.String()
 		if wgMsg != "" {
-			serverLogger.Warn("waitgroups waiting", "msgs", wgMsg)
+			clog.Warn(ctx, "waitgroups waiting", "msgs", wgMsg)
```
Comment (sb10) [inline cmd/manager.go:750]: This should be ctxf.

File: cmd/manager.go [hunk 19]
```diff
@@ -735,23 +747,24 @@ func startJQ(postCreation []byte) {
...
 			wgDebug.Reset()
 		}
-		serverLogger.Crit("deadlock", "err", deadlockBuf.String())
+
+		clog.Crit(ctx, "deadlock", "err", deadlockBuf.String())
```
Comment (sb10) [inline cmd/manager.go:754]: This should be ctxf.

File: cmd/manager.go [hunk 20]
```diff
@@ -773,15 +785,17 @@ func startJQ(postCreation []byte) {
...
+	// to get final message below
+	clog.AddHandler(fileHandler)
 
 	// block forever while the jobqueue does its work
 	err = server.Block()
 
 	wgMsg := wgDebug.String()
 	if wgMsg != "" {
-		serverLogger.Warn("waitgroups waiting", "msgs", wgMsg)
+		clog.Warn(ctx, "waitgroups waiting", "msgs", wgMsg)
```
Comment (sb10) [inline cmd/manager.go:798]: This should be ctxf.

File: cmd/manager.go [hunk 21]
```diff
@@ -804,7 +818,7 @@ func startJQ(postCreation []byte) {
 // so that the next time the manager is started it will create a new token.
 // For un-clean exits of the manager, we should keep the token so the manager
 // re-uses it, allowing any runners to reconnect.
-func deleteToken() {
+func deleteToken(ctx context.Context) {
```
Comment (sb10) [inline cmd/manager.go:821]: ctx not used?

File: cmd/mod.go [hunk 22]
```diff
@@ -124,7 +126,7 @@ new internal ids is printed.`,
 		}()
 
 		// get the job(s) user wishes to modify
-		jobs := getJobs(jq, jobqueue.JobStateDeletable, cmdAll, 0, false, false)
+		jobs := getJobs(ctx, jq, jobqueue.JobStateDeletable, cmdAll, 0, false, false)
```
Comment (sb10) [inline cmd/mod.go:129]: Probably a few of these private methods in cmd package no longer need ctx; go through and check them all.

File: cmd/mount.go [hunk 23]
```diff
@@ -264,7 +266,7 @@ func init() {
 
 // mountParse takes possible json string or simple string (as per `wr mount -h`)
 // and parses exactly 1 of them to a MountConfig for each mount defined.
-func mountParse(jsonString, simpleString string) jobqueue.MountConfigs {
+func mountParse(ctx context.Context, jsonString, simpleString string) jobqueue.MountConfigs {
```
Comment (sb10) [inline cmd/mount.go:269]: ctx not used?

File: cmd/status.go [hunk 24]
```diff
@@ -473,7 +475,7 @@ func countGetJobArgs() int {
 	return set
 }
 
-func getJobs(jq *jobqueue.Client, cmdState jobqueue.JobState, all bool, statusLimit int, showStd, showEnv bool) []*jobqueue.Job {
+func getJobs(ctx context.Context, jq *jobqueue.Client, cmdState jobqueue.JobState, all bool, statusLimit int, showStd, showEnv bool) []*jobqueue.Job {
```
Comment (sb10) [inline cmd/status.go:478]: Probably ctx isn't needed?

File: jobqueue/server.go [hunk 25]
```diff
@@ -1223,10 +1223,10 @@ func (s *Server) createQueue() {
 	// package will only call this once at a time, so we don't need to worry
 	// about locking across the whole function.
 	q.SetReadyAddedCallback(func(queuename string, allitemdata []interface{}) {
-		defer internal.LogPanic(s.Logger, "jobqueue ready added callback", true)
+		defer internal.LogPanic(ctx, "jobqueue ready added callback", true)
 
-		s.Debug("rac started")
-		defer s.Debug("rac finished")
+		clog.Error(ctx, "rac started")
```
Comment (sb10) [inline jobqueue/server.go:1228]: If it was s.Debug, then it should be clog.Debug. Double-check all replacements in all files; there's at least a few more of these mis-translations below.

**Review by sb10 — CHANGES_REQUESTED** [review]
Some fixes needed. I've only looked at changes since the last time; once these fixes are in I'll need to do a from-scratch review of everything and do some manual testing as well.

File: cmd/manager.go (hunk 21 above)
Comment (ac55-sanger) [reply cmd/manager.go]: Yes, all these ctx were mainly to pass to die,info,warn methods of root file and now not needed. 
I have updated all these and also removed other unnecessary ctx.

File: cmd/manager.go [hunk 26]
```diff
@@ -605,7 +601,9 @@ func logStarted(ctx context.Context, s *jobqueue.ServerInfo, token []byte) {
 	}
 }
 
-func startJQ(ctx context.Context, postCreation []byte) {
+func startJQ(postCreation []byte) {
+	ctx := context.Background()
```
Comment (sb10) [inline cmd/manager.go:605]: If this is only used in the clog.ContextWithFileHandler call, move it inside there to avoid confusion.

**Review by sb10 — CHANGES_REQUESTED** [review]
Thanks. A little change and then I need to do the full review which will take a while.

File: cmd/manager.go (hunk 26 above)
Comment (ac55-sanger) [reply cmd/manager.go:605]: It's also passed to cloud.New (line 719) that's why I kept it as a variable

Comment (sb10) [reply cmd/manager.go:605]: Ok, fair enough. I'll do the full review now.

File: limiter/limiter.go [hunk 27]
```diff
@@ -1,4 +1,4 @@
-// Copyright © 2019 Genome Research Limited
+// Copyright © 2019, 2021 Genome Research Limited
```
Comment (sb10) [inline limiter/limiter.go:1]: 2019-2021

File: main_test.go [hunk 28]
```diff
@@ -1,4 +1,4 @@
-// Copyright © 2020 Genome Research Limited
+// Copyright © 2020, 2021 Genome Research Limited
```
Comment (sb10) [inline main_test.go:1]: `git rm` the file entirely

File: queue/queue.go [hunk 29]
```diff
@@ -1,4 +1,4 @@
-// Copyright © 2016-2019 Genome Research Limited
+// Copyright © 2016-2019, 2021 Genome Research Limited
```
Comment (sb10) [inline queue/queue.go:1]: 2016-2021

File: queue/queue.go (hunk 6 above)
Comment (sb10) [reply queue/queue.go:1212]: Still not sure why this takes a context

File: queue/queue_test.go [hunk 30]
```diff
@@ -1,4 +1,4 @@
-// Copyright © 2016, 2018, 2019 Genome Research Limited
+// Copyright © 2016, 2018, 2019, 2021 Genome Research Limited
```
Comment (sb10) [inline queue/queue_test.go:1]: 2016-2021

File: queue/subqueue.go [hunk 31]
```diff
@@ -1,4 +1,4 @@
-// Copyright © 2016, 2019 Genome Research Limited
+// Copyright © 2016, 2019, 2021 Genome Research Limited
```
Comment (sb10) [inline queue/subqueue.go:1]: 2016, 2017, 2019-2021

**Review by sb10 — CHANGES_REQUESTED** [review]
A bunch of minor changes please.

**Review by sb10 — CHANGES_REQUESTED** [review]
There are some undesirable differences in the logging when comparing develop and config branches.

$ wr cloud deploy --debug
$ echo "sleep 1" | wr add -i sleep
$ echo "sleep 2" | wr add -i sleep --cpus 2
$ wr cloud teardown
[compare ~/.wr_development/log.openstack for the 2 branches]

- For the add, normal CLI WARN logs shouldn't have caller info.

develop:
msg="server allocate" server=localhost [...] caller=server.go:313
msg="server create attempted" cloud=openstack took=701.785877ms id=ad697f45-2460-4996-baaa-f838a5f397c8 worked=true caller=openstack.go:854
msg="server allocate" server=ad697f45-2460-4996-baaa-f838a5f397c8 cores=2.000 [...] caller=server.go:
313

config:
msg="server allocate" serverid= schedulertype=openstack callvalue=18e73ff030a7f720 [...] caller=clog.go:212
msg="server create attempted" schedulertype=openstack cloudtype=openstack callvalue=1dfa92b98beea439 serverflavor=s2.small id=70ddda20-01ce-4e53-a2 [...] caller=clog.go:212
0e-a57a771b15c6 worked=true caller=clog.go:212
msg="server allocate" serverid=70ddda20-01ce-4e53-a20e-a57a771b15c6 schedulertype=openstack callvalue=75c3ad5df52f1ccf cores=2.000 [...] caller=clog.go:212

- caller values need to report higher up the call stack.
- Duplicated schedulertype and cloudtype.
- No serverid for localhost.

Check there are no other significant differences, and check the logs are the same without --debug when the server issues a warning or worse.

**Comment by sb10** [comment]
The recent change to add localhost broke running jobs on localhost:

```
t=2021-08-16T09:34:11+0000 lvl=dbug msg="server went bad" schedulertype=openstack server=localhost caller=scheduler/openstack.go:1252
[...]
t=2021-08-16T09:34:11+0000 lvl=dbug msg="ran cmd" schedulertype=openstack callvalue=4622384674a232fd cmd="/tmp/wr runner -s '1024:60:1:0:6a9df116ae2cb47fa4adcb70142346f2' --deployment development --server
 '192.168.67.111:46407' --domain localhost -r 1 -m 61 --debug" call=4622384674a232fd caller=scheduler/local.go:665
t=2021-08-16T09:34:11+0000 lvl=dbug msg="runCmd error" schedulertype=openstack callvalue=4622384674a232fd err="no available server" caller=scheduler/local.go:693
```

Take another look and find an alternative way of getting it to report localhost in the logs.

**Comment by ac55-sanger** [comment]
> The recent change to add localhost broke running jobs on localhost:
> 
> ```
> t=2021-08-16T09:34:11+0000 lvl=dbug msg="server went bad" schedulertype=openstack server=localhost caller=scheduler/openstack.go:1252
> [...]
> t=2021-08-16T09:34:11+0000 lvl=dbug msg="ran cmd" schedulertype=openstack callvalue=4622384674a232fd cmd="/tmp/wr runner -s '1024:60:1:0:6a9df116ae2cb47fa4adcb70142346f2' --deployment development --server
>  '192.168.67.111:46407' --domain localhost -r 1 -m 61 --debug" call=4622384674a232fd caller=scheduler/local.go:665
> t=2021-08-16T09:34:11+0000 lvl=dbug msg="runCmd error" schedulertype=openstack callvalue=4622384674a232fd err="no available server" caller=scheduler/local.go:693
> ```
> 
> Take another look and find an alternative way of getting it to report localhost in the logs.

I tested the multiple job submissions and it

File: cloud/server.go [hunk 32]
```diff
@@ -293,7 +294,12 @@ func (s *Server) Matches(os string, script []byte, configFiles string, flavor *F
 // there was enough space. Returns true if there was enough space and the
 // allocation occurred.
 func (s *Server) Allocate(ctx context.Context, cores float64, ramMB, diskGB int) bool {
-	ctx = clog.ContextWithServerID(ctx, s.ID)
+	if s.Name == localhostName {
```
Comment (sb10) [inline cloud/server.go:297]: This shouldn't just be for Allocate and Release, and the code block shouldn't be duplicated. All current calls to clog.ContextWithServerID should instead be replaced with a call to a new method s.ContextWithServerID, which calls clog.ContextWithServerID(ctx, id), where id is localhostName if s.Name is localhostName.

**Review by sb10 — CHANGES_REQUESTED** [review]
